pyomo/contrib/sensitivity_toolbox/sens.py

```diff
@@ -7,216 +7,684 @@
+        components = (components,)
+    for comp in components:
+        if comp.is_indexed():
+            for idx in sorted_robust(comp):
+                yield idx, comp[idx]
+        else:
+            yield _NotAnIndex, comp
+
+def sensitivity_calculation(method, instance, paramList, perturbList,
+         cloneModel=True, tee=False, keepfiles=False, solver_options=None):
+    sens = SensitivityInterface(instance, clone_model=cloneModel)
+    sens.setup_sensitivity(paramList)
+
+    m = sens.model_instance
+
+    if method == 'kaug':
+        kaug = SolverFactory('k_aug', solver_io='nl')
+        dotsens = SolverFactory('dot_sens', solver_io='nl')
+        ipopt = SolverFactory('ipopt', solver_io='nl')
+
+        ipopt.solve(m, tee=tee)
+        m.ipopt_zL_in.update(m.ipopt_zL_out)  #: important!
+        m.ipopt_zU_in.update(m.ipopt_zU_out)  #: important!    
+
+        kaug.options['dsdp_mode'] = ""  #: sensitivity mode!
+        kaug.solve(m, tee=tee)
+        m.write('col_row.nl', format='nl', io_options={'symbolic_solver_labels':True})
+
+    sens.perturb_parameters(perturbList)
+
+    if method == 'sipopt':
+        ipopt_sens = SolverFactory('ipopt_sens', solver_io='nl')
+        ipopt_sens.options['run_sens'] = 'yes'
+
+        # Send the model to the ipopt_sens and collect the solution
+        results = ipopt_sens.solve(m, keepfiles=keepfiles, tee=tee)
+
+    elif method == 'kaug':
+        dotsens.options["dsdp_mode"] = ""
+        dotsens.solve(m, tee=tee)
+        try:
+            os.makedirs("dsdp")
+        except FileExistsError:
+            # directory already exists
+            pass
+        try:
+            shutil.move("dsdp_in_.in","./dsdp/")
+            shutil.move("col_row.nl","./dsdp/")
+            shutil.move("col_row.col","./dsdp/")
+            shutil.move("col_row.row","./dsdp/")
+            shutil.move("conorder.txt","./dsdp/")
+            shutil.move("delta_p.out","./dsdp/")
+            shutil.move("dot_out.out","./dsdp/")
+            shutil.move("timings_dot_driver_dsdp.txt", "./dsdp/")
+            shutil.move("timings_k_aug_dsdp.txt", "./dsdp/")
+        except OSError:
+            pass
+
+    return m
+
+def get_dsdp(model, theta_names, theta, var_dic={},tee=False, solver_options=None):
+    """This function calculates gradient vector of the (decision variables, parameters)
+        with respect to the paramerters (theta_names).
+    e.g) min f:  p1*x1+ p2*(x2^2) + p1*p2
+         s.t  c1: x1 + x2 = p1
+              c2: x2 + x3 = p2
+              0 <= x1, x2, x3 <= 10
+              p1 = 10
+              p2 = 5
+    the function retuns dx/dp and dp/dp, and colum orders.
     Parameters
     ----------
-    instance: ConcreteModel
-        pyomo model object
+    model: Pyomo ConcreteModel
+        model should includes an objective function
+    theta_names: list of strings
+        List of Var names
+    theta: dict
+        Estimated parameters e.g) from parmest
```

Comment:
Is there  any reason these are passed in as a list of names and a dict of values, instead of just a list of components (Vars or Params)? Do the names typically come from serialized data?

Reply:
I made this function with parmest. So, I followed the same format with parmest. We might remove `theta_names`, and define them in the functions with `theta_names=theta.keys() `.